packages/utils/telemetry-utils/src/config.ts

```diff
@@ -0,0 +1,307 @@
+    if (isPrimitiveType(outputType)) {
+        return { value: output, type: outputType };
+    }
+
+    if (Array.isArray(output)) {
+        const firstType = typeof output[0];
+        if (!isPrimitiveType(firstType)) {
+            return defaultReturn;
+        }
+        for (const v of output) {
+            if (typeof v !== firstType) {
+                return defaultReturn;
+            }
+        }
+        return { value: output, type: `${firstType}[]` as ConfigTypeStrings };
+    }
+
+    return defaultReturn;
+}
+
+const safeSessionStorage = (): Storage | undefined => {
+    try {
+        return sessionStorage !== null ? sessionStorage : undefined;
+    } catch { return undefined; }
+};
+
+interface ConfigCacheEntry extends Partial<ConfigTypeStringToType> {
+    readonly raw: ConfigTypes;
+    readonly name: string;
+}
+
+/**
+ * Implementation of {@link IConfigProvider} which contains nested {@link IConfigProviderBase} instances
+ */
+export class ConfigProvider implements IConfigProvider {
+    private readonly configCache = new Map<string, ConfigCacheEntry>();
+    private readonly orderedBaseProviders: (IConfigProviderBase | undefined)[];
+    private static readonly logger = DebugLogger.create("fluid:telemetry:configProvider");
+
+    static create(orderedBaseProviders: (IConfigProviderBase | ITelemetryBaseLogger | undefined)[],
+    ): IConfigProvider {
+        const filteredProviders: (IConfigProviderBase | undefined)[] = [];
+        for (const maybeProvider of orderedBaseProviders) {
+            if (maybeProvider !== undefined) {
+                if (isConfigProviderBase(maybeProvider)) {
+                    filteredProviders.push(maybeProvider);
+                } else {
+                    if (loggerIsMonitoringContext(maybeProvider)) {
+                        filteredProviders.push(maybeProvider.config);
+                    }
+                }
+            }
+        }
+        return new ConfigProvider(filteredProviders);
+    }
+
+    private constructor(
+        orderedBaseProviders: (IConfigProviderBase | undefined)[],
+    ) {
+        this.orderedBaseProviders = [];
+        const knownProviders = new Set<IConfigProviderBase>();
+        const candidateProviders = [...orderedBaseProviders];
+        while (candidateProviders.length > 0) {
+            const baseProvider = candidateProviders.shift()!;
+            if (baseProvider !== undefined
+                && isConfigProviderBase(baseProvider)
+                && !knownProviders.has(baseProvider)
+            ) {
+                knownProviders.add(baseProvider);
+                if (baseProvider instanceof ConfigProvider) {
+                    candidateProviders.push(...baseProvider.orderedBaseProviders);
+                } else {
+                    this.orderedBaseProviders.push(baseProvider);
+                }
+            }
+        }
+    }
+    getBoolean<TD extends boolean | undefined>(name: string, defaultValue: TD): TD | boolean {
```

Comment:
@andre4i so i played around with typing for the default value and landed here,  we need to generic that infers from the parameter otherwise consumer always need to deal with undefined, even if the return can't be undefined because default is a real value. given this, i'm not sure default value is even worth it, and i'm not sure its worth it to introduce this complicated typeing.

example without the generic 
``
getBoolean(name: string, defaultValue: boolean | undefined): undefined | boolean 

// the typing here will be undefined | boolean eventhough it can't be undefined because my default is false
// so now i need to check, or use  ! or ?? anyway
const myconfig = getBoolean("name", false);
``

Reply:
i think i' prefer not to support a default, then consumers just do
``` typescript
const myconfig = getBoolean("name") ?? false;
```
and the typing is right without any generics or complicated typing

Reply:
Yes, looking closely at how this API would be used currently in the codebase, I agree with you. We do stuff like `const myconfig = getBoolean("name", false) === true` anyway which makes the default value kind of cumbersome.